### PR TITLE
feat: Transaction creation options are supported by all transactions

### PR DESCRIPTION
**Pull request by amanda-tarafa**
feat: Support excluding a transaction from change streams

### CODE REVIEW

**Comment by amanda-tarafa** [comment]
@jskeet as always, one commit at a time.

- All top commits, except for the last one, are adding full support for transaction creation options. Review these carefully, in particular:
  - Naming
  - I hacked away at the overload explosion that we had been accumulating, in particular for the SpannerConnectio.BeginTransaction set of methods. The idea is that, if we want to offer helpers for building different types of transaction, we do that via SpannerTransactionCreationOptions. Still it might be useful to leave a few of the more commons overloads.
  - The last commint add support for the new transaction creation option, "exclude transaction from change stream". It's very easy to add those type of options now, and that commit is in this PR to demonstrate the benefits of the approach.
  - There's still similar work to be done for adding full support for the transaction options that may be set after transaction creation.

**Comment by amanda-tarafa** [comment]
Another thing to note is that individual commits won't build on their own, I'll squash them before merging, but for reviews it's better as is.

**Comment by amanda-tarafa** [comment]
(I'm looking at PR errors, but all are because C# language version)

**Comment by amanda-tarafa** [comment]
All green and ready for review now.

**Review by jskeet — APPROVED** [review]
Various nits and questions - but broadly looking good :)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransactionCreationOptions.cs [hunk 1]
```diff
@@ -92,16 +91,38 @@ public sealed class SpannerTransactionCreationOptions
...
-    internal TransactionOptions TransactionOptios => IsPartitionedDml
-        ? PartitionedDmlTransactionOptions
-        : TransactionMode == TransactionMode.ReadWrite
-        ? ReadWriteTransactionOptions
-        : TimestampBound?.ToTransactionOptions();
+    internal TransactionOptions TransactionOptios
+    {
+        get
+        {
+            var options = IsPartitionedDml
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerTransactionCreationOptions.cs:112]: Yep, done, but now in a method.

**Review by amanda-tarafa — COMMENTED** [review]
All comments addressed and changes pushed in ordered "review" commits. Thanks!

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs [hunk 2]
```diff
@@ -1063,6 +1063,121 @@ public async Task MaxCommitDelay_SetOnCommand_UnsetOnAmbientTransaction_CommandI
                 Arg.Any<CallSettings>());
         }
 
+        [Fact]
+        public async Task TransactionOptions_Propagates_ExplicitTransaction()
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:1067]: I think they're fine. It's just an inherently detailed area.

**Comment by jskeet** [comment]
In SpannerCommandTests.cs, you've got this reply to a comment of mine - I can't see my comment now, or reply to yours:

> SpannerDataReader does not implement IAsyncDisposable, or did you mean something else?

I suspect I meant changing this:

```csharp
using (var reader = await command.ExecuteReaderAsync())
{
    ...
}
```

into:

```csharp
using var reader = await command.ExecuteReaderAsync();
...
```

**Review by jskeet — APPROVED** [review]
Changes all look good. I hope I've checked all the responses - GitHub hasn't been terribly helpful on some of them.

**Comment by amanda-tarafa** [comment]
> In SpannerCommandTests.cs, you've got this reply to a comment of mine - I can't see my comment now, or reply to yours:
> 
> > SpannerDataReader does not implement IAsyncDisposable, or did you mean something else?
> 
> I suspect I meant changing this:
> 
> ```cs
> using (var reader = await command.ExecuteReaderAsync())
> {
>     ...
> }
> ```
> 
> into:
> 
> ```cs
> using var reader = await command.ExecuteReaderAsync();
> ...
> ```

Ah OK, yes, I'll do that, I though you meant `await using...`

**Review by amanda-tarafa — COMMENTED** [review]
Latest comments addressed: indent and simplified usings.

I'm running all tests locally and will merge on green.Thanks!

**Pull request merged**